jobrunner/agent/main.py

```diff
@@ -0,0 +1,241 @@
+import logging
+import sys
+import time
+
+from opentelemetry import trace
+
+from jobrunner import config
+from jobrunner.agent import task_api
+from jobrunner.executors import get_executor_api
+from jobrunner.job_executor import ExecutorAPI, ExecutorState, JobDefinition, JobStatus
+from jobrunner.lib.log_utils import configure_logging, set_log_context
+from jobrunner.models import StatusCode
+from jobrunner.queries import get_flag_value
+
+
+log = logging.getLogger(__name__)
+tracer = trace.get_tracer("loop")
+
+EXECUTOR_RETRIES = {}
+
+
+class RetriesExceeded(Exception):
+    pass
+
+
+class InvalidTransition(Exception):
+    pass
+
+
+class ExecutorError(Exception):
+    pass
+
+
+def main(exit_callback=lambda _: False):  # pragma: no cover
+    log.info("jobrunner.run loop started")
+    api = get_executor_api()
+
+    while True:
+        with tracer.start_as_current_span("LOOP", attributes={"loop": True}):
+            active_tasks = handle_tasks(api)
+
+        if exit_callback(active_tasks):
+            break
+
+        time.sleep(config.JOB_LOOP_INTERVAL)
+
+
+def handle_tasks(api: ExecutorAPI | None):
+    active_tasks = task_api.get_active_tasks()
+
+    handled_tasks = []
+
+    for task in active_tasks:
+        # `set_log_context` ensures that all log messages triggered anywhere
+        # further down the stack will have `job` set on them
+        with set_log_context(task=task.id):
```

Comment:
Do we mean task id or job id now? When we wrote this, we were assuming they were the same

Reply:
Good point.

Reply:
Hmm, actually, at this point, we don't know if the task is job related or not

So I think task.id is correct.

In handle_run_job_task, we can *add* the job id to the log context